libraries/psio/include/psio/reflect.hpp

```diff
@@ -0,0 +1,760 @@
+#define PSIO_MEMBER_POINTER(STRUCT, members)                                          \
+   BOOST_PP_IIF(BOOST_PP_CHECK_EMPTY(members), PSIO_EMPTY, PSIO_MEMBER_POINTER_IMPL2) \
+   (STRUCT, members)
+
+#define PSIO_PROXY_DATA(r, STRUCT, i, elem)                                                \
+   psio::member_proxy<i, psio::hash_name(BOOST_PP_STRINGIZE(PSIO_GET_IDENT(elem))),        \
+                                         &STRUCT::PSIO_GET_IDENT(elem), ProxyObject>       \
+                      PSIO_GET_IDENT(elem)()                                               \
+   {                                                                                       \
+      return _psio_proxy_obj;                                                              \
+   }                                                                                       \
+   psio::member_proxy<i, psio::hash_name(BOOST_PP_STRINGIZE(PSIO_GET_IDENT(elem))),        \
+                                         &STRUCT::PSIO_GET_IDENT(elem), const ProxyObject> \
+                      PSIO_GET_IDENT(elem)() const                                         \
+   {                                                                                       \
+      return _psio_proxy_obj;                                                              \
+   }
+
+#define PSIO_PROXY_METHOD(r, STRUCT, i, elem)                                           \
+   template <typename... Args>                                                          \
+   auto PSIO_GET_IDENT(elem)(Args... args)                                              \
+   {                                                                                    \
+      psio::member_proxy<i, psio::hash_name(BOOST_PP_STRINGIZE(PSIO_GET_IDENT(elem))),  \
+                                            &STRUCT::PSIO_GET_IDENT(elem), ProxyObject> \
+                             m(_psio_proxy_obj);                                        \
+      return m.call(std::forward<decltype(args)>(args)...);                             \
+   }
+
+/**
+ * PSIO_REFLECT(<struct>, <member or base spec>...)
+ * Each parameter may be one of the following:
+ *    * ident                          non-static data member
+ *    * definitionWillNotChange()      the struct's definition will not change in the future. This saves 2 bytes from the fracpack representation.
+ *    * allowHashedMethods()           allow hashed method identifiers
+ *    * base(ident)                    base class (not implemented)
+ *    * method(ident, arg1, ...)       method
+ *    * numbered(int, ident)           non-static data member with field number
+ */
+#define PSIO_REFLECT(STRUCT, ...)                                                                 \
+   PSIO_REFLECT_TYPENAME(STRUCT)                                                                  \
+   struct reflect_impl_##STRUCT                                                                   \
```

Comment:
I would prefix all internal symbol names that are inserted in the user's namespace with `psio_` to avoid conflicts.